invoices/sql_migration_test.go

```diff
@@ -0,0 +1,434 @@
+	// we do in the real world, but rather we strive to test the migration
+	// code paths. Therefore some data might be inconsistent or not make
+	// much sense.
+	invoice := &Invoice{
+		Memo:           []byte(randString(10)),
+		PaymentRequest: []byte(randString(MaxPaymentRequestSize)),
+		CreationDate:   created,
+		SettleDate:     settled,
+		Terms:          terms,
+		// For our tests we don't care about the add index since it
+		// won't be migrated as all SQL invoices have a SQL generated
+		// ID instead of the add index.
+		AddIndex:    0,
+		SettleIndex: settleIndex,
+		State:       state,
+		AMPState:    make(map[SetID]InvoiceStateAMP),
+		HodlInvoice: rand.Intn(2) == 1,
+	}
+
+	// Now generate some HTLCs for the invoice.
+	invoice.Htlcs = make(map[models.CircuitKey]*InvoiceHTLC)
+
+	if invoice.IsAMP() {
+		generateAMPHtlcs(t, invoice)
+	} else {
+		generateInvoiceHTLCs(invoice)
+	}
+
+	for _, htlc := range invoice.Htlcs {
+		if htlc.State == HtlcStateSettled {
+			invoice.AmtPaid += htlc.Amt
+		}
+	}
+
+	return invoice
+}
+
+// randString generates a random string of length n.
+func randString(n int) string {
+	const letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = letters[rand.Intn(len(letters))]
+	}
+
+	return string(b)
+}
+
+// randTimeBetween generates a random time between min and max.
+func randTimeBetween(min, max time.Time) time.Time {
+	var timeZones = []*time.Location{
+		time.UTC,
+		time.FixedZone("EST", -5*3600),
+		time.FixedZone("MST", -7*3600),
+		time.FixedZone("PST", -8*3600),
+		time.FixedZone("CEST", 2*3600),
+	}
+
+	// Ensure max is after min
+	if max.Before(min) {
+		min, max = max, min
+	}
+
+	// Calculate the range in nanoseconds
+	duration := max.Sub(min)
+	randDuration := time.Duration(rand.Int63n(duration.Nanoseconds()))
+
+	// Generate the random time
+	randomTime := min.Add(randDuration)
+
+	// Assign a random time zone
+	randomTimeZone := timeZones[rand.Intn(len(timeZones))]
+
+	// Return the time in the random time zone
+	return randomTime.In(randomTimeZone)
+}
+
+// randTime generates a random time between 2009 and 2140.
+func randTime() time.Time {
+	min := time.Date(2009, 1, 3, 0, 0, 0, 0, time.UTC)
+	max := time.Date(2140, 1, 1, 0, 0, 0, 1000, time.UTC)
+
+	return randTimeBetween(min, max)
+}
+
+func randInvoiceTime(invoice *Invoice) time.Time {
+	return randTimeBetween(
+		invoice.CreationDate,
+		invoice.CreationDate.Add(invoice.Terms.Expiry),
+	)
+}
+
+func randHTLC(invoice *Invoice, amt lnwire.MilliSatoshi) (models.CircuitKey,
+	*InvoiceHTLC) {
+
+	htlc := &InvoiceHTLC{
+		Amt:          amt,
+		AcceptHeight: uint32(rand.Intn(999) + 1),
+		AcceptTime:   randInvoiceTime(invoice),
+		Expiry:       uint32(rand.Intn(999) + 1),
+	}
+	if invoice.Terms.Features.HasFeature(lnwire.MPPRequired) {
+		htlc.MppTotalAmt = invoice.Terms.Value
+	}
+
+	// For simplicity we just reflect the invoice state in the HTLC.
```

Comment:
and not in the invoice itself ?

Reply:
Not sure I understand what you meant, could you pls clarify?